contributors/design-proposals/preserve-order-in-strategic-merge-patch.md

```diff
@@ -0,0 +1,437 @@
+
+The new version patch is always a superset of the old version patch.
+The new patch has one additional parallel list which will be dropped by the old server.
+
+The new directive is optional.
+Patch requests without the directive will retain the behavior from previous releases.
+This will ensure backward compatibility with old clients.
+
+This change is fully backward compatible.
+
+### kubectl
+If an old kubectl sends a old patch to a new server, the server will not preserve the order which behave as an old server.
+
+If a new kubectl sends a new patch to an old server, the server doesn't recognise the parallel list and will drop it.
+So it will behave the same as before.
+
+## Example
+
+### List of Maps
+
+We take environment variables as an example.
+Environment variables is a list of maps with merge patch strategy.
+
+Suppose we define a list of environment variables and we call them
+the original environment variables:
+
+```yaml
+env:
+- name: ENV1
+  value: foo
+- name: ENV2
+  value: bar
+- name: ENV3
+  value: baz
+```
+
+Then the server appends two environment variables and reorder the list:
+
+```yaml
+env:
+- name: ENV2
+  value: bar
+- name: ENV5
+  value: server-added-2
+- name: ENV1
+  value: foo
+- name: ENV3
+  value: baz
+- name: ENV4
+  value: server-added-1
+```
+
+Then the user wants to change it from the original to the following using `kubectl apply`:
+
+```yaml
+env:
+- name: ENV1
+  value: foo
+- name: ENV2
+  value: bar
+- name: ENV6
+  value: new-env
+```
+
+The old patch without parallel list will looks like:
+
+```yaml
+env:
+- name: ENV3
+  $patch: delete
+- name: ENV6
+  value: new-env
+```
+
+The new patch will looks like below. It is the
+
+```yaml
+$setElementOrder/env:
+- name: ENV1
+- name: ENV2
+- name: ENV6
+env:
+- name: ENV3
+  $patch: delete
+- name: ENV6
+  value: new-env
+```
+
+After server applying the patch:
+
+```yaml
+env:
+- name: ENV1
+  value: foo
+- name: ENV2
+  value: bar
+- name: ENV6
+  value: new-env
+- name: ENV5
+  value: server-added-2
+- name: ENV4
+  value: server-added-1
+```
+
+### List of Primitives
+
+We take finalizers as an example.
+finalizers is a list of strings.
+
+Suppose we define a list of finalizers and we call them
+the original finalizers:
+
+```yaml
+finalizers:
+- a
+- b
+- c
+```
+
+Then the server appends two finalizers and reorder the list:
+
+```yaml
+finalizers:
+- b
+- e
+- a
+- c
+- d
+```
+
+Then the user wants to change it from the original to the following using `kubectl apply`:
+
+```yaml
+finalizers:
+- a
+- b
+- f
+```
+
+The old patch without parallel list will looks like:
+
+```yaml
+$deleteFromPrimitiveList/finalizers:
+- c
+finalizers:
+- f
+```
+
+The new patch will looks like below. It is the
+
+```yaml
+$setElementOrder/finalizers:
+- a
+- b
+- f
+$deleteFromPrimitiveList/finalizers:
+- c
+finalizers:
+- f
+```
+
+After server applying the patch:
+
+```yaml
+finalizers:
+- a
+- b
+- f
+- e
+- d
+```
+
+
+# Alternative Considered
```

Comment:
Thinking on this more, if we did that, we might end up creating elements that we just want to order - e.g. if we want to modify "A", and send "A, B, C" as the main list to keep ordering, "C" would get recreated if it was deleted or missing, which we would not want.